OptimizeRasters.py

```diff
@@ -1426,24 +1427,56 @@ def pyramidLevels(self):
         return _steps
 
     def _call_external(self, args):
-        p = subprocess.Popen(' '.join(args), shell=True,
-                             stdout=subprocess.PIPE)
-        message = '/'
-        CSIZE_PREFIX = b'Size is'
-        while (message):
-            message = p.stdout.readline()
-            if (message):
-                _strip = message.strip()
-                if (_strip.find(CSIZE_PREFIX) != -1):
-                    wh = _strip.split(CSIZE_PREFIX)
-                    if (len(wh) > 1):
-                        wh = wh[1].split(b',')
-                        if (self.CW in self._propertyNames):
-                            self.width = int(wh[0].strip())
-                        if (self.CH in self._propertyNames):
-                            self.height = int(wh[1].strip())
-                self._GDALInfo.append(_strip)
-        return len(self._GDALInfo) > 0
+        """
+        Typically runs gdalinfo.exe with the supplied arguments
+        Appends GDAL stdout to messages
+        If return code for output of subprocess.run is not 0 this indicates an error, so the code waits and retries
+        """
+
+        # TODO compare log output size vs Esri\OptimizeRasters
+
+        args = ' '.join(args)
+        self.message(rf"INFO: GDALInfo._call_external args = {args}")
+
+        max_number_attempts=5
+        retry_delay_seconds=1
+
+        for attempt_number in range(1, max_number_attempts + 1):
+            completedProcess = subprocess.run(args, capture_output=True)
+            if completedProcess.returncode == 0:
+                break
+
+            self.message(rf"WARN: Attempt {attempt_number}: subprocess.run() return code indicates error. stderr = {completedProcess.stderr}")
+
+            if (attempt_number) <= max_number_attempts:
+                self.message(rf"INFO: Retrying in {retry_delay_seconds} seconds")
+                time.sleep(retry_delay_seconds)
+                continue
+
+            raise subprocess.CalledProcessError(completedProcess.returncode, args, completedProcess.stderr)
+
+        self.message(rf"INFO: GDALInfo._call_external succeeded after {attempt_number} attempts")
+        
+        # INFO: stdout from gdalinfo prints a lot of unwanted image info, so we don't add it to messages
+
+        stdout_decoded = completedProcess.stdout.decode()
+        CSIZE_regex_match = 'Size is (\d*), (\d*)'
+        width_and_height_matches = re.search(CSIZE_regex_match, stdout_decoded)
+
+        # I don't really like this but I'm trying to avoid breaking original functionality
+        if (self.CW in self._propertyNames):
+            if not width_and_height_matches:
+                raise Exception(rf"ERROR: Failed to find regex match {CSIZE_regex_match} in stdout")
+            self.width = int(width_and_height_matches.groups()[0])
+            self.message(rf"INFO: self.width set to {self.width}")
+
```

Comment:
I think you could remove 1472-1475 and the code would be the same

Reply:
Most likely yes, but I've tried to avoid any changes to the existing functionality which does check for self.CW and self.CH, so I'll keep this in.